### PR TITLE
Introduce dockertest to integration tests

### LINKED ISSUE #177: Execute integration tests in CI
## Is your feature request related to a problem? Please describe

Run integration tests (e.g. depending on external components like kafka , db etc ... ) in an automated way as part of the normal build.

## Describe the solution

Investigate usage of https://github.com/ory/dockertest to spin up external components

### PR DESCRIPTION
**Pull request by drakos74**
Signed-off-by: Vangelis Katikaridis <v.katikaridis@thebeat.co>

## Which problem is this PR solving?

Closes #177.

## Short description of the changes

- Make use of dockertest to manage docker containers from within the go test code.
- use the testify suite wrapper for the test to manage setup and teardown actions

## TODO :
- add a special tag as part of normal ci pipeline
i.e. -tags integration_unit_test or -tags component_test or something ...

### CODE REVIEW

**Review by mantzas — CHANGES_REQUESTED** [review]
you have to vendor in order to make the ci work.

My first impression is that it strays away from idiomatic go testing.
I had more something in mind using TestMain to:
- Setup go containers needed for the tests
- run std idiomatic go tests as tey existed previoulsy
- teardown the go containers.

**Comment by drakos74** [comment]
> you have to vendor in order to make the ci work.
> 
> My first impression is that it strays away from idiomatic go testing.
> I had more something in mind using TestMain to:
> 
> * Setup go containers needed for the tests
> * run std idiomatic go tests as tey existed previoulsy
> * teardown the go containers.

Good point for the vendors, missed that . Thanks!

Regarding the tests, that was the idea, to keep the idiomatic go testing pattern.
The test conditions and logic remains the same, as well as the test code.

The only functional difference now, is that the test needs to wait for the container to start.
So i thought it might be cleaner to use the testify "TestSuite" tools.

Having said that, we can do without the 'suite' ... 
We would need to run a method at the start of the tests to setup the containers :
`setupContainersSync(...)`
and at the end another one to stop them :
`tearDownContainersSync(...)`

What do you think ?

**Comment by mantzas** [comment]
> > you have to vendor in order to make the ci work.
> > My first impression is that it strays away from idiomatic go testing.
> > I had more something in mind using TestMain to:
> > 
> > * Setup go containers needed for the tests
> > * run std idiomatic go tests as tey existed previoulsy
> > * teardown the go containers.
> 
> Good point for the vendors, missed that . Thanks!
> 
> Regarding the tests, that was the idea, to keep the idiomatic go testing pattern.
> The test conditions and logic remains the same, as well as the test code.
> 
> The only functional difference now, is that the test needs to wait for the container to start.
> So i thought it might be cleaner to use the testify "TestSuite" tools.
> 
> Having said that, we can do without the 'suite' ...
> We would need to run a method at the start of the tests to setup the containers :
> `setupContainersSync(...)`
> and at the end another one to stop them :
> `tearDownContainersSync(...)`
> 
> What do you think ?

I believe that would be nice and idiomatic. Setup and teardown in TestMain.
Go 1.14 will also introduce teardown per test so we might eventually use this also to clean up between tests.

File: trace/sql/integration_test.go [hunk 1]
```diff
@@ -6,92 +6,102 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"os"
 	"testing"
 	"time"
 
+	"github.com/beatlabs/patron/log"
+
+	"github.com/jmoiron/sqlx"
```
Comment (mantzas) [inline trace/sql/integration_test.go]: Let's avoid adding unneeded packages

File: trace/sql/integration_test.go [hunk 2]
```diff
@@ -6,92 +6,102 @@ import (
...
-
-	s.db.SetConnMaxLifetime(time.Minute)
-	s.db.SetMaxIdleConns(10)
-	s.db.SetMaxOpenConns(10)
-
+	return nil
 }
 
-func (s *SQLTestSuite) TailLogs(containerID string, out io.Writer) {
+func TailLogs(d *dockerRuntime, containerID string, out io.Writer) {
```
Comment (mantzas) [inline trace/sql/integration_test.go]: as the linter suggests we can remove this.

File: trace/sql/integration_test.go [hunk 3]
```diff
@@ -6,92 +6,102 @@ import (
...
-	ctx  context.Context
 }
 
-func TestSQLTestSuite(t *testing.T) {
-	suite.Run(t, new(SQLTestSuite))
+func TestMain(m *testing.M) {
+
+	d := dockerRuntime{}
+
+	err := startUpContainerSync(&d)
```
Comment (mantzas) [inline trace/sql/integration_test.go]: We could add startup and teardon to the dockerRuntime struct.

File: trace/sql/integration_test.go [hunk 4]
```diff
@@ -6,92 +6,102 @@ import (
...
-	//s.TailLogs(s.sql.Container.ID, os.Stdout)
+	// optionally print the container logs in stdout
+	//TailLogs(d, d.sql.Container.ID, os.Stdout)
 
 	// wait until the container is ready
-	err = s.pool.Retry(func() error {
-		connString := fmt.Sprintf("%s:%s@(%s:%s)/%s?parseTime=true&multiStatements=true", DBUsername, DBPassword, DBHost, DBPort, DBSchema)
-		db, err := Open("mysql", connString)
+	err = d.pool.Retry(func() error {
+		db, err := sqlx.Open("mysql", fmt.Sprintf(connectionFormat, dbUsername, dbPassword, dbHost, dbPort, dbSchema))
```
Comment (mantzas) [inline trace/sql/integration_test.go]: Let's use the std sql package to test and avoid another package dependecy

**Review by mantzas — CHANGES_REQUESTED** [review]
Except for the a few improvements, looks really nice...

File: trace/sql/integration_test.go (hunk 2 above)
Comment (drakos74) [reply trace/sql/integration_test.go]: i ve just enabled it. So now we see the logs from the container. Would that be ok ?

File: trace/sql/integration_test.go [hunk 5]
```diff
@@ -288,3 +330,72 @@ func assertSpan(t *testing.T, sp *mocktracer.MockSpan, opName, statement string)
...
+		db, err := sql.Open("mysql", fmt.Sprintf(connectionFormat, dbUsername, dbPassword, dbHost, dbPort, dbSchema))
+		if err != nil {
+			// container not ready ... return error to try again
+			return err
+		}
+		return db.Ping()
+	})
+}
+
+func (d *dockerRuntime) TailLogs(containerID string, out io.Writer) {
```
Comment (mantzas) [inline trace/sql/integration_test.go]:
```suggestion
func (d *dockerRuntime) tailLogs(containerID string, out io.Writer) {
```

**Review by mantzas — DISMISSED** [review]

**Review by mantzas — DISMISSED** [review]

**Comment by mantzas** [comment]
I really like that we can have now everything inside our codebase and not scattered around in env files, docker compose, etc...
I really like it a lot!
Kudos

**Review by moukoublen — DISMISSED** [review]

**Review by moukoublen — APPROVED** [review]

**Review by mantzas — APPROVED** [review]

**Comment by mantzas** [comment]
@drakos74 let's wait also for the other curators got get some majority into this one... @sotirispl @cpapidas @fpapadopou

**Pull request merged**